unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/StepsController/ActionController.cs

```diff
@@ -0,0 +1,149 @@
+            {
+                UndoCurrentAction();
+                if (currentStepIndex > 0)
+                    currentStepIndex--;
+            }
+            else
+            {
+                RedoCurrentAction();
+                if (currentStepIndex + 1 < actionsMade.Count)
+                    currentStepIndex++;
+            }
+        }
+
+    }
+
+    public void TryToRedoAction()
+    {
+        if (currentStepIndex < actionsMade.Count)
+        {
+            if (currentStepIndex + 1 < actionsMade.Count)
+            {
+                if(currentStepIndex == 0 && actionsMade[currentStepIndex].isDone)
+                    currentStepIndex++;
+                else if(currentStepIndex != 0)
+                    currentStepIndex++;
+                RedoCurrentAction();
+            }
+        }
+    }
+
+    public void TryToUndoAction()
+    {
+        if (currentStepIndex >= 0 && actionsMade.Count > 0)
+        {
+            UndoCurrentAction();
+            if (currentStepIndex > 0)
+                currentStepIndex--;
+        }
+    }
+
+    public void AddAction(BuildModeAction action)
+    {
+        bool removedActions = false;
+        if (currentStepIndex < actionsMade.Count-1)
+        {
+            actionsMade.RemoveRange(currentStepIndex, actionsMade.Count - currentStepIndex);
+            removedActions = true;
+        }
+        actionsMade.Add(action);
+    
+        currentStepIndex = actionsMade.Count-1;
+        if (removedActions && actionListview != null)
+            actionListview.SetContent(actionsMade);
+        action.OnApplyValue += ApplyAction;
+    }
+
+    void ApplyAction(DecentralandEntity entityToApply, object value, BuildModeAction.ActionType actionType)
+    {
+        switch (actionType)
+        {
+            case ActionType.MOVE:
+                Vector3 convertedPosition = (Vector3)value;
+                entityToApply.gameObject.transform.position = convertedPosition;
+                break;
+            case ActionType.ROTATE:
+                Vector3 convertedAngles = (Vector3)value;
+                entityToApply.gameObject.transform.eulerAngles = convertedAngles;
+                break;
+            case ActionType.SCALE:
+                Vector3 convertedScale = (Vector3)value;
+                Transform parent = entityToApply.gameObject.transform.parent;
+
+                entityToApply.gameObject.transform.localScale = new Vector3(convertedScale.x / parent.localScale.x, convertedScale.y / parent.localScale.y, convertedScale.z / parent.localScale.z);
+                break;
+            case ActionType.CREATED:
+                string entityString = (string)value;
+                if (entityString.Length < 255)
+                {
+                    buildModeController.DeleteEntity((string)value);
+                }
+                else
+                {
+                    buildModeController.CreateEntityFromJSON((string)value);
+                }
```

Comment:
How does relate the `entityString.Length` to the decision of deleting/creating an entity? Furthermore, why an action `ActionType.CREATED` would delete an entity? Is it a fallback or an invalid entity?

Reply:
This needs to be refactored with the implementation with kernel. At the moment to create it, I didn't know how this approach should work so I just create a JSON with the content of the entity knowing that it will be refactored in the future before entering production.

For the name:
Each action can be undone, redone. This action in particular is created when an entity is created. If you Undo this action, the created entity should be destroyed. As stated above, this will be reworked to not depends on the length of the string

Reply:
Sounds good!